MediaRanker - completed wave 3

vikshab opened the pull request:
I did some refactoring on duplicated logic, DRYed my code 

I haven't used shared_examples to test controllers. All tests are passed and I have 100% coverage. Still feels like I might have missed something to test. Would like to have comments on my partials for edit and new forms, particularly how to make less locals, the lines looks long because of them

vikshab committed:
Gemset

vikshab committed:
ruby version

vikshab committed:
WIP working add, delete, view all buttons

vikshab committed:
Models created

vikshab committed:
WIP

vikshab committed:
Added rspec-rails and simplecov gems

vikshab committed:
Generated rspec:install and rspec:model for all models

vikshab committed:
Added require simplecove into spec_helper

vikshab committed:
Upvote button ready for movies

vikshab committed:
Movies - ready

vikshab committed:
Changed redirect_to in destroy action

vikshab committed:
Added custom routes for upvote - books and albums

vikshab committed:
Added upvoke action to albums controller

vikshab committed:
Completed views for albums and books

vikshab committed:
WIP - scope in models

vikshab committed:
Scope best created for 3 models

vikshab committed:
WIP

vikshab committed:
WIP

vikshab committed:
WIP

vikshab committed:
Movie controller edit and update methods updated

vikshab committed:
Fixed create and update methods for books and albums controllers

vikshab committed:
Green tests for all models

vikshab committed:
Refacting code

vikshab committed:
100% test coverage for movie controller

vikshab committed:
100% test coverage for books and albums controller

vikshab committed:
100% test coverage for users controller

vikshab committed:
Added bootstrap -sass gem

vikshab committed:
Changes application.css to application.scss and added @import

vikshab committed:
Added bootstrap-sprokets to javascripts/application.js

vikshab committed:
Header partial move to layout

vikshab committed:
Partials for new and edit forms are ready for movies/books/albums

vikshab committed:
Partials ready for movies

vikshab committed:
Partials ready for books

vikshab committed:
Partials ready for albums

vikshab committed:
53 green tests and fixed some extra spaces between lines

vikshab committed:
Refactoring code

vikshab committed:
Added a background picture to the header and fixed buttons layout

vikshab committed:
Fixed table header position

vikshab committed:
Run test, 100% coverage

vikshab committed:
Deleted a line of extra code in update methods in all three controllers

vikshab committed:
fixed the updated method test in controllers

vikshab committed:
New seeds data

vikshab committed:
Changed index method in controllers - order('ranking DESC')

vikshab committed:
Fixed director in seeds

vikshab committed:
Created shared examples for models

vikshab committed:
shared_examples wip

vikshab committed:
Changed update method in controllers

vikshab committed:
Wave 3 completed

kariabancroft commented on .gitignore:
```diff
@@ -1,35 +1,17 @@
-.powenv
+# Ignore all logfiles and tempfiles.
+/log/*
+!/log/.keep
```
Recommend adding all .keep files to your gitignore

kariabancroft commented on app/controllers/albums_controller.rb:
```diff
@@ -0,0 +1,63 @@
+class AlbumsController < ApplicationController
+
+  def index
+    @albums = Album.order('ranking DESC')
```
It would be better to encapsulate the `ranking DESC` logic within the model rather than in the controller

vikshab replied:
moved this logic to the model - `scope :desc_order` (for all models )

kariabancroft commented on app/controllers/albums_controller.rb:
```diff
@@ -0,0 +1,63 @@
+  end
+
+  def edit
+    @album = Album.find(params[:id])
```
You use this line multiple times to retrieve your model item. I'd recommend implementing a before_action that applies to each of these specific actions

vikshab replied:
Moved `@album = Album.find(params[:id])` into private method `find_album` and created `before-action :find_album, only: [:show, :edit, :update, :destroy :upvote]`` Did the same in other controllers

kariabancroft commented on app/controllers/albums_controller.rb:
```diff
@@ -0,0 +1,63 @@
+  def update
+    @album = Album.find(params[:id])
+
+    @album.update(create_params[:album])
```
the `update` method will actually do the save so you don't need to call `save` again

kariabancroft commented on app/controllers/albums_controller.rb:
```diff
@@ -0,0 +1,63 @@
+  # ______________________________________________________________________
+  private
+
+  def create_params
```
You use this method for more than just model creation so I'd recommend re-naming this to be more generic

vikshab replied:
Changed `create_params` to `album_params`. Did the same changes in other controllers

kariabancroft commented on app/controllers/albums_controller.rb:
```diff
@@ -0,0 +1,63 @@
+  end
+
+  def destroy
+    show
```
I like that you use the show logic here to retrieve your album, but this would be better suited in a before action or separate private method call because when the destroy action is called, it is not showing anything so calling this method is a bit confusing

vikshab replied:
I used `before_action`

kariabancroft commented on app/controllers/books_controller.rb:
```diff
@@ -0,0 +1,63 @@
+class BooksController < ApplicationController
```
Same comments from Albums controller apply here

vikshab replied:
Fixed

kariabancroft commented on app/controllers/movies_controller.rb:
```diff
@@ -0,0 +1,63 @@
+class MoviesController < ApplicationController
```
Same comments from Albums controller apply here

vikshab replied:
Fixed

kariabancroft commented on app/models/album.rb:
```diff
@@ -0,0 +1,7 @@
+  validates :name, presence: true
+
+  # Scope_______________________________________________________________________
+  scope :best, -> { order('ranking DESC').limit(10) }
```
It might be better to have this scope (and the other models) accept a parameter and then use one single constant within the UsersController index action

vikshab replied:
I changed the `scope :best, -> { order('ranking DESC').limit(10) }` to `scope :best, -> { order('ranking DESC') }` for all the models, created a constant `LIMIT` in `users_controller`, so 
the index method looks like `@movie = Movie.desc_order.limit(LIMIT)`

kariabancroft commented on app/views/albums/show.html.erb:
```diff
@@ -0,0 +1,3 @@
+<%= render partial: "shared/show_content", locals: {name: @album.name, by: @album.artist, ranking: @album.ranking, description: @album.description } %>
+
+<%= render partial: "shared/buttons_footer", locals: {upvote_name_path: upvote_album_path(@album.id), edit_name: @album.name, edit_name_path: edit_album_path(@album.id), name_path: album_path(@album.id), view_name: "Albums", all_names_path: albums_path} %>
```
I would recommend putting the locals on separate lines so like:

```
<%= render partial: "shared/buttons_footer", 
                       locals: {upvote_name_path: upvote_album_path(@album.id),         
                                   edit_name: @album.name, 
                                   edit_name_path: edit_album_path(@album.id), 
                                   name_path: album_path(@album.id), 
                                   view_name: "Albums", 
                                   all_names_path: albums_path} %>
```

vikshab replied:
Fixed

kariabancroft commented on app/views/albums/index.html.erb:
```diff
@@ -0,0 +1,11 @@
+<table class="table table-hover">
```
I would recommend including this whole table structure within the index_content shared partial and utilize the list of each model type as a local, rather than duplicating the table structure within each index view

kariabancroft commented on app/views/shared/_new_form.html.erb:
```diff
@@ -0,0 +1,14 @@
+<%= form_for subject, html: {class: "form"}, url: path do |f| %>
```
Since this form is very close to the edit for, I'd recommend trying to create _one_ shared form view that can handle both edit and new

vikshab replied:
Created just one partial `_form.html.erb` for new and edit

kariabancroft commented on config/routes.rb:
```diff
@@ -0,0 +1,62 @@
+  patch 'books/:id/upvote'  => "books#upvote",  as: "upvote_book"
+  patch 'albums/:id/upvote' => "albums#upvote", as: "upvote_album"
+
+  # Example of regular route:
```
Once you feel comfortable with your routes, you should delete the default routes that are included in these comments

vikshab replied:
Did you mean all the commented out routes examples?

kariabancroft commented on coverage/.resultset.json:
```diff
@@ -0,0 +1,1312 @@
+{
+  "RSpec": {
+    "coverage": {
```
All coverage information should be included in your gitignore file

vikshab replied:
Added to .gitignore `coverage/`

kariabancroft commented on spec/controllers/albums_controller_spec.rb:
```diff
@@ -0,0 +1,224 @@
+
+# INDEX ACTION__________________________________________________________________
+
+  describe "GET #index" do
```
All of these index-based tests are totally independent of the controller, so this would be a great start in attempting to implement a shared example for all controllers.

kariabancroft commented on spec/controllers/albums_controller_spec.rb:
```diff
@@ -0,0 +1,224 @@
+  describe "GET #show" do
+    before :each do
+      @album = Album.create(album_params[:album])
+      @album.ranking = 0
```
This would be a great thing to put as a default value within your model, then `ranking` would always be set on every create even if you didn't explicitly set it on the model

kariabancroft commented on spec/controllers/albums_controller_spec.rb:
```diff
@@ -0,0 +1,224 @@
+      @album = Album.create(album_params[:album])
+    end
+
+    let(:album_params) do
```
It looks like this would be a great place to attempt to refactor to use factorygirl in the future, since each time i've seen these album params, which create a model object, the params are the same (or very very similar).

kariabancroft commented on spec/support/shared_examples_for_models.rb:
```diff
@@ -0,0 +1,33 @@
+require 'rails_helper'
+
+RSpec.shared_examples "a collection" do
```
You did a nice job with this shared example for your models.

kariabancroft commented:
You did a really nice job on this. There are a few things that I would attempt to clean up in the controllers, but overall they look good. I think the # of the locals you use with your partials is okay, I'd just separate it out into separate lines to abide by the style guide and therefore make it easier to read.

vikshab committed:
moved this logic to the model - scope :desc_order (for all models )
Moved @album = Album.find(params[:id]) into private method find_album and created before-action :find_album, only: [:show, :edit, :update, :destroy :upvote]` Did the same in other controllers

vikshab committed:
Changed create_params to album_params, movie_params and book_params

vikshab committed:
I changed the scope :best, -> { order('ranking DESC').limit(10) } to scope :best, -> { order('ranking DESC') } for all the models, created a constant LIMIT in users_controller, so
the index method looks like @movie = Movie.desc_order.limit(LIMIT)

vikshab committed:
Put the locals on separate lines for all the view pages

vikshab committed:
Deleted partial for edit form, instead I created just one partial _form.html.erb for edit and new form
for album, movie and book models

vikshab committed:
Aded to gitignore coverage/

vikshab committed:
Fixed intend

kariabancroft commented:
Nice job refactoring your controllers and scopes!
